docs/design/draft/initial-design.md

```diff
@@ -0,0 +1,379 @@
+    - Find VolumeSnapshotContent based on Status.BoundVolumeSnapshotContentName
+    - If snapshotContent.Status is empty or Status.SnapshotHandle is empty, keep waiting
+    - Any other Status fields to check?
+
+  - Create new PVC with VolumeSnapshot DataSource, using VolumeSnapshot from above
+    - request.Storage from source PVC Status.Capacity.Storage
+    - StorageClassName from source PVC
+    - update `DataMoverBackup` `Status` with Spec PVC fields and
+      additional metadata to include in backup (storageClassName,
+      accessModes, storage)
+
+- Once the PVCs are created, create a Job resource with an image which includes the data mover plugin:
+  - Where is the line between Data Mover core and the plugin?
+    The Job image may be mostly plugin with core elements dealing with input metadata.
+  - New PVCs created above are all mounted in the job pod
+  - `DataMoverBackup` reference passed to Job pod (how? env var?)
+
+### Restores
+
+For each new `DataMoverRestore`:
+- For each PVC to be restored:
+  - Create new PVC using Namespace, Name, Storageclass, size, AccessMode from DataMoverRestore
+  - Fill in PVC-related Status fields fields
+  - Update DataMoverRestore CR Status with restic snapshot ID (this
+    may need to go into a `config` Spec field since it's plugin-specific)
+
+- Once the empty PVCs are created, create a Job resource with an image which includes the data mover plugin:
+  - New PVCs created above are all mounted in the job pod
+  - `DataMoverRestore` reference passed to Job pod (how? env var?)
+
+## Data Mover Job
+
+This is a short-lived pod created as a `Job` resource that's run once
+per `DataMoverBackup` or `DataMoverRestore` resource. We can probably
+use the same image for both operations, but it's still an open
+question. There's no reason we can't use separate images if it results
+in a cleaner implementation. The plugin could be built into the image,
+or the plugin could be included as an initContainer similar to Velero
+plugins.
+
+### Backups
+
+`DataMoverBackup` is passed into the pod. For each PVC in `Status`
+(all should be mounted), perform a restic backup from the mounted
+PVC into the referenced `StorageLocation`. Store appropriate per-PVC
+Status metadata into the `StorageLocation` as well:
+- Restic Snapshot ID
+- Original PVC namespaced name
+- PVC capacity
+- StorageClassName
+
+The metadata and all snapshots should be stored under a top level
+directory which uniquely identifies the `DataMoverBackup` (name, UID,
+etc.)
+
+### Restores
+
+`DataMoverRestore` is passed into the pod. For each PVC in `Status`
+(all should be mounted), perform a restic restore into the mounted PVC.
+
+## Other Open issues to address:
+
+- Possible alternate design approach: Instead of watching for
+  `DataMoverBackup` resources, the Data Mover controller could watch
+  CSI snapshots instead. It depends on whether we want a
+  "backup-centric" approach, or a "PV/snapshot-centric" approach. The
```

Comment:
Data mover watching for CSI snapshots might be useful in situations where a backup application does not depend on velero (atleast for backing up PV data). In situations where the backup application wants to invoke CSI snapshots by itself (instead of triggering a backup operation to Velero which in turn would take CSI snapshots), if the data mover keeps watching for CSI snapshots, it can get active and start backing up PV data once it finds a new CSI snapshot. 

Otherwise, the backup applications which does not depend on velero for backup of PV data might find it bit difficult to move the PV data to the storage location, as data mover would be watching for "DataMoverBackup" resources and they would not be created if the backup application takes CSI snapshot directly.

I am still getting familiar with velero and other kubernetes concepts. Sorry if I have missed something obvious and the above point mentioned does not hold correct.

Regards,
Raghavendra
My 2 c

Reply:
Actually, nothing in this design so far depends on Velero. While a velero backup with CSI plugin might be the primary way of producing CSI snapshots, manual VolumeSnapshot creation or some other tool would have the same effect. Everything in this document that mentions Backups and Restores is referring to the Data Mover proposed CRDs: DataMoverBackup and DataMoverRestore. See the above "Data Mover CRDs" section. The question here is more whether the Data Mover should back up CSI snapshots in response to an explicit call for it to do so (by creating a DataMoverBackup resource) or whether this should happen automatically and implicitly every time a CSI snapshot is created, either by Velero or by some other means. The overall design here assumes an explicit DataMoverBackup resource. The comments below in this section address an alternate approach (in response to a suggestion from John), with a listing of some of the potential drawbacks of that approach.